discord/components/file_upload.py

```diff
@@ -0,0 +1,116 @@
+FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+DEALINGS IN THE SOFTWARE.
+"""
+
+from __future__ import annotations
+
+from typing import TYPE_CHECKING, ClassVar, Literal
+
+from typing_extensions import override
+
+from ..enums import ComponentType
+from ..types.component_types import FileUpload as FileUploadPayload
+from .component import Component, ModalComponentMixin
+
+if TYPE_CHECKING:
+    from typing_extensions import Self
+
+
+class FileUpload(ModalComponentMixin[FileUploadPayload], Component[FileUploadPayload]):
+    """Represents a File Upload Component.
+
+    This component displays a file upload box in a :class:`Modal`.
+
+    This inherits from :class:`Component`.
+
+    .. versionadded:: 3.0
+
+    Attributes
+    ----------
+    type: Literal[:data:`ComponentType.file_upload`]
+        The type of component.
+    custom_id: :class:`str`
+        The custom ID of the file upload component that gets received during an interaction.
+    min_values: :class:`int`
+        The minimum number of files that must be uploaded.
+    max_values: :class:`int`
+        The maximum number of files that can be uploaded.
+    required: :class:`bool`
+        Whether the file upload is required to submit the modal.
+    id: :class:`int` | :data:`None`
+        The section's ID.
+
+    Parameters
+    ----------
+    id:
+        The component's ID. If not provided by the user, it is set sequentially by Discord.
+        The ID `0` is treated as if no ID was provided.
+    """
+
+    __slots__: tuple[str, ...] = (
+        "file",
+        "spoiler",
+        "name",
+        "size",
+    )
+
+    __repr_info__: ClassVar[tuple[str, ...]] = __slots__
+    versions: tuple[int, ...] = (2,)
+    type: Literal[ComponentType.file_upload] = ComponentType.file_upload  # pyright: ignore[reportIncompatibleVariableOverride]
+
+    def __init__(
+        self,
+        custom_id: str,
+        id: int | None = None,
+        min_values: int = 1,
+        max_values: int = 1,
+        required: bool = True,
+    ) -> None:
+        self.custom_id: str = custom_id
+        self.min_values: int = min_values
+        self.max_values: int = max_values
+        self.required: bool = required
+        super().__init__(id=id)
+
+    @classmethod
+    @override
+    def from_payload(cls, payload: FileUploadPayload) -> Self:
+        return cls(
+            custom_id=payload["custom_id"],
+            id=payload["id"],
+            min_values=payload.get("min_values", 1),
+            max_values=payload.get("max_values", 1),
+            required=payload.get("required", True),
+        )
+
+    @override
+    def to_dict(self, modal: bool = True) -> FileUploadPayload:
+        payload: FileUploadPayload = {
+            "type": int(self.type),
+            "custom_id": self.custom_id,
+            "min_values": self.min_values,
+            "max_values": self.max_values,
+            "required": self.required,
+            "id": self.id,
```

Comment:
id is typed as int | None however FileUploadPayload id cant be none as its a literal 19, maybe there is some thing to look for here (same for from payload etc)

Reply:
THose are all to be ignored, id None can be sent to discord and will be handled as 0, it just will always be an int when recieved

Reply:
i dont get the point of the id as from the payload it can only be 19 since its the file upload one
So why are we letting the user change it ? 
(mb just realize that type is the 19, id is something else...)
And so why are we not putting it into the payload as notrequired ? 
```
Could not access item in TypedDict
  "id" is not a required key in "FileUpload", so access may result in runtime
  ```

Reply:
Because it is always there when we get the payload

Reply:
then why not add it to the payload ? 
<img width="357" height="135" alt="Screenshot 2025-11-10 at 14 51 14" src="https://github.com/user-attachments/assets/c6fc2e5f-b67a-4c8d-9f9f-c7cb29761d81" />

> Because it is always there when we get the payload

Reply:
It is on BaseComponent

Reply:
but shouldnt it be override since it will always be and so required? 

Im thinking about something like this 
```py
class FileUpload(BaseComponent):
    id: int # Will be 0 if not provided (and not required)
    type: Literal[19]
    custom_id: str
    min_values: NotRequired[int]
    max_values: NotRequired[int]
    required: NotRequired[bool]
```

Reply:
if you really want to be pedantic due to the typing, just
```suggestion
            "id": self.id or 0,
```